generalize dims as X/Y when lat/lon or rlat/rlon

vindelico opened the pull request:
<!-- Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [ ] (If applicable) Documentation has been added / updated (for bug fixes / features).
- [ ] (If applicable) Tests have been added.
- [x] This PR does not seem to break the templates.
- [x] HISTORY.rst has been updated (with summary of main changes).
 - [ ] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.

### What kind of change does this PR introduce?

* spatial dimensions can be generalized as X/Y when rechunking and will be mapped to rlon/rlat or lon/lat accordingly.
* The yml file can contain

rechunk:
   time: -1
   X: 100
   Y: 100

instead of 

rechunk:
   time: -1
   rlat: 100
   rlon: 100

or 

rechunk:
   time: -1
   lat: 100
   lon: 100

This allows to handle grids on different projections together. Example: RDRSv2 (rlat/rlon) and ERA5-Land (lat/lon); different CORDEX-NA RCMs; 

### Does this PR introduce a breaking change?

No.

### Other information:

The duplicated code preparing for rechunking from `save_to_netcdf()` and `save_to_zarr()` was refactored into a private function `_rechunk_for_saving()`

aulemahal commented on xscen/io.py:
```diff
@@ -518,6 +488,29 @@ def coerce_attrs(attrs):
+            rechunk_dims = rechunk.copy()
+
+        # get actual axes labels
+        if {"X", "Y"}.issubset(rechunk_dims):
+            rechunk_dims[ds.cf.axes["X"][0]] = rechunk_dims.pop("X")
+            rechunk_dims[ds.cf.axes["Y"][0]] = rechunk_dims.pop("Y")
```
```suggestion
        if "X"  in rechunk_dims and "X" not in ds.dims:
            rechunk_dims[ds.cf.axes["X"][0]] = rechunk_dims.pop("X")
        if "Y"  in rechunk_dims and "Y" not in ds.dims:
            rechunk_dims[ds.cf.axes["Y"][0]] = rechunk_dims.pop("Y")
```
To allow chunking along only one of the dimensions and to do your trick only when there's not a real "X" dimension.

vindelico replied:
Good catch! I like it.

aulemahal reviewed (approved):
Nice addition! Thanks!

aulemahal commented on xscen/io.py:
```diff
@@ -518,6 +488,29 @@ def coerce_attrs(attrs):
 
 
+def _rechunk_for_saving(ds, rechunk):
+    """Do chunking before saving to .zarr or .nc, generalized as Y/X for different axes lat/lon, rlat/rlon"""
```
```suggestion
    """Rechunk before saving to .zarr or .nc, generalized as Y/X for different axes lat/lon, rlat/rlon."""
```
To satisfy the nitpicky pre-commit check.

juliettelavoie commented on HISTORY.rst:
```diff
@@ -16,6 +16,7 @@ New features and enhancements
 * New function `get_warming_level` to search within the IPCC CMIP global temperatures CSV without requiring data. (:issue:`208`, :pull:`210`).
 * File re-structuration from catalogs with ``xscen.catutils.build_path``. (:pull:`205`).
 * New scripting functions `save_and_update` and `move_and_delete`. (:pull:`214`).
+* spatial dimensions can be generalized as X/Y when rechunking and will be mapped to rlon/rlat or lon/lat accordingly.
```
```suggestion
* Spatial dimensions can be generalized as X/Y when rechunking and will be mapped to rlon/rlat or lon/lat accordingly. (:pull: `221`).
```

juliettelavoie reviewed (commented):
Nice! Are you able to add a test ? You could create a `test/test_io.py` file with the same structure as the other test files with a `test_rechunk_for_saving` function.

RondeauG commented:
Looks good! I would add a mention of this both in the Docstrings for `save_to_zarr` and `save_to_netcdf` (probably where `rechunk` is defined), as well as in the documentation: https://xscen.readthedocs.io/en/latest/notebooks/2_getting_started.html#Saving-files-to-disk

vindelico commented on HISTORY.rst:
```diff
@@ -16,6 +16,7 @@ New features and enhancements
 * New function `get_warming_level` to search within the IPCC CMIP global temperatures CSV without requiring data. (:issue:`208`, :pull:`210`).
 * File re-structuration from catalogs with ``xscen.catutils.build_path``. (:pull:`205`).
 * New scripting functions `save_and_update` and `move_and_delete`. (:pull:`214`).
+* Spatial dimensions can be generalized as X/Y when rechunking and will be mapped to rlon/rlat or lon/lat accordingly. (:pull: `221`).
```
```suggestion
* Spatial dimensions can be generalized as X/Y when rechunking and will be mapped to rlon/rlat or lon/lat accordingly. (:pull:`221`).
```

vindelico commented:
> Looks good! I would add a mention of this both in the Docstrings for `save_to_zarr` and `save_to_netcdf` (probably where `rechunk` is defined), as well as in the documentation: https://xscen.readthedocs.io/en/latest/notebooks/2_getting_started.html#Saving-files-to-disk

I changed the docstrings and and added a line about this in the notebook.

vindelico commented:
> Nice! Are you able to add a test ? You could create a `test/test_io.py` file with the same structure as the other test files with a `test_rechunk_for_saving` function.

test/test_io.py created.
I opted for four tests, to check if lat/lon, rlat/rlon are rechunked and when either is rechunked with X/Y as dims.

juliettelavoie commented on tests/test_io.py:
```diff
@@ -0,0 +1,101 @@
+                [[14.7, 10.1, 17.0], [3.4, 23.5, 15.1]],
+                [[8.5, 12.4, 25.8], [14.5, 5.8, 15.3]],
+            ]
+            # pr = [[[9.81, 0.00, 8.88],
```
remove code that was commented out.

vindelico replied:
Done.

juliettelavoie commented on tests/test_io.py:
```diff
@@ -0,0 +1,101 @@
+import pandas as pd
+import pytest
+import xarray as xr
+import xclim.xclim.sdba.measures
```
```suggestion
import xclim.sdba.measures
```

juliettelavoie replied:
actually, I think this i never used. I would just remove the line.

vindelico replied:
Good catch, removed.

juliettelavoie commented on xscen/io.py:
```diff
@@ -518,6 +492,30 @@ def coerce_attrs(attrs):
             raise
 
 
+def _rechunk_for_saving(ds, rechunk):
```
I think you need to add it to the `__all__` at the begining of the file to be able to import it correctly in the tests.

juliettelavoie replied:
I would also consider removing the "_", as users might want to use this function directly ?

vindelico replied:
All done, thank you!

vindelico commented on HISTORY.rst:
```diff
@@ -16,6 +16,7 @@ New features and enhancements
 * New function `get_warming_level` to search within the IPCC CMIP global temperatures CSV without requiring data. (:issue:`208`, :pull:`210`).
 * File re-structuration from catalogs with ``xscen.catutils.build_path``. (:pull:`205`).
 * New scripting functions `save_and_update` and `move_and_delete`. (:pull:`214`).
+* Spatial dimensions can be generalized as X/Y when rechunking and will be mapped to rlon/rlat or lon/lat accordingly. (:pull:`221`).
```
Done.

RondeauG commented on tests/conftest.py:
```diff
@@ -43,3 +47,119 @@ def samplecat():
+
+
+@pytest.fixture
+def datablock_3d():
```
@aulemahal I implemented this to support the need to test 3D data in `xscen`. Thoughts?
